File: September Challenge/Week 2/Java/Day_8/SumofRootToLeafBinaryNumbers.java
// Problem : https://leetcode.com/problems/sum-of-root-to-leaf-binary-numbers/
// Solution:

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int tot;
    public int sumRootToLeaf(TreeNode root) {
        tot=0;
        calc(root,0);
        return tot;
    }
    
    private void calc(TreeNode node, int sumsf){
        if(node==null) return;
        int sum = 2*sumsf + node.val;
        if(node.left==null && node.right==null){
            tot+=sum;
            return;
        }
        calc(node.left,sum);
        calc(node.right,sum);
    }
}

File: September Challenge/Week 2/Cpp/DAY_8/SumOfRootToLeafBinaryNumbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findSum(TreeNode* root ,int value1){
        if(root == nullptr){
            return 0;
        }
        
        
        value1 = (value1*2) + root->val;
        if(root->left==NULL && root->right==NULL){
            return value1;
        }
        return findSum(root->left , value1) + findSum(root->right , value1);
    }
    int sumRootToLeaf(TreeNode* root) {
        
        int sum = findSum(root , 0);
        return sum;
    }
};
